LSPS5/README.md

```diff
@@ -0,0 +1,623 @@
+
+The client can register a webhook under a specific name, list the
+names of registered webhooks, and remove named webhooks.
+
+* `lsps5.set_webhook` - Accepts an `app_name` string and `webhook`
+  URL, which adds a new webhook under that name (or replaces the
+  existing webhook under that name).
+* `lsps5.list_webhooks` - Returns a list of `app_name` strings
+  that have currently-registered webhooks.
+* `lsps5.remove_webhook` - Removes the webhook with a given
+  `app_name`.
+
+If the LSP needs some action from the client, but the client is
+not currently connected to the LSP, the LSP contacts all
+registered webhooks.
+
+The LSP supports some maximum number of webhooks.
+
+> **Rationale** Each webhook not only represents some information
+> that the LSP must store in persistent storage, but whenever the
+> LSP needs to wake up the client, the LSP needs to contact
+> **all** webhooks, thus each webhook is also a potential increase
+> in network bandwidth use.
+
+#### `lsps5.set_webhook`
+
+The client uses the `lsps5.set_webhook` call to specify the URI
+that the LSP should contact in order to send a push notification
+to the client user.
+The `lsps5.set_webhook` API takes the parameters:
+
+```JSON
+{
+    "app_name": "My LSPS-Compliant Lightning Client",
+    "webhook": "https://www.example.org/push?l=1234567890abcdefghijklmnopqrstuv&c=best"
+}
+```
+
+`app_name` is a required string, containing a human-readable
+UTF-8 string that gives a name to the webhook.
+The `app_name` can be up to 64 bytes in UTF-8 format, with `\`
+escape codes counting as multiple bytes (e.g. `\a` is two bytes,
+`\000` is 4 bytes), but not counting the `"` double quote
+delimiters.
+
+> **Rationale** Backslash escapes may represent characters that
+> are problematic in some contexts, such as `\000` when using C
+> string libraries, and implementations may store the string
+> with the backslash escapes as-is.
+>
+> The use of backslash escapes is dubious in both `app_name`
+> and `webhook` contexts (the latter should use URI-encoding),
+> but is allowed since JSON allows it.
+>
+> The `app_name` is specified to be human-readable, to allow
+> the creation of a user interface that can manage multiple
+> webhooks.
+
+`webhook` is a required string, containing the URI of the
+webhook that the LSP can use to push a notification to the
+client.
+The `webhook` can be up to 1024 bytes in UTF-8 format, with
+the same rules as for `app_name`.
+
+The `webhook` MUST have one of the following protocols:
+
+* `https`
+
+The client:
+
+* MUST use a HTTPS-protocol URI to the notification delivery
+  service.
+* SHOULD encode the following data in any part of the URI
+  (hostname, path, or `GET` parameters):
+  * The LSP node ID.
+  * Some authorization to allow the given LSP to wake the
+    client, ideally a proof that the client has a channel
+    or other relationship with the LSP.
+  * Some way to identify the target device the client is
+    running on.
```

Comment:
The LSP ?Node ID is needed by the client + notification delivery server, not the LSP, and we are using the URL as a pass-through by which the client can give any information needed by the client or notification delivery server.

* The notification delivery service needs to know the LSP node ID so it can validate the signature in the header.  Note that it cannot be simply sent by the LSP in the same webhook notification `POST` request, because if some hacker figures out the schema of the notification delivery service or otherwise acquires the webhook of a client (say an LSP leak which acquires the webhooks database but not the LSP node private key) then the hacker can generate fake LSP node IDs to spam the client with.  So the client needs to (1) provide the LSP node ID to the notification delivery service and (2) give an authorization to the notification delivery service that "yes, this LSP node ID is good, accept this LSP node ID but reject everyone else".
* If the client has multiple LSPs, it needs to know the LSP node ID of the source of the notification.